libdrgn/kallsyms.c

```diff
@@ -0,0 +1,611 @@
+{
+	struct drgn_error *err;
+	const size_t token_index_size = (UINT8_MAX + 1) * sizeof(uint16_t);
+	uint64_t last_token;
+	size_t token_table_size, names_idx;
+	char data;
+	uint8_t len_u8;
+	int len;
+	bool bswap;
+
+	err = drgn_program_bswap(prog, &bswap);
+	if (err)
+		return err;
+
+	// Read num_syms from vmcore (bswap is done for us already)
+	err = drgn_program_read_u32(prog,
+				    loc->kallsyms_num_syms,
+				    false, &kr->num_syms);
+	if (err)
+		return err;
+
+	// Read the constant-sized token_index table (256 entries)
+	kr->token_index = malloc(token_index_size);
+	if (!kr->token_index)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->token_index,
+				       loc->kallsyms_token_index,
+				       token_index_size, false);
+	if (err)
+		return err;
+	if (bswap)
+		for (size_t i = 0; i < kr->num_syms; i++)
+			kr->token_index[i] = bswap_16(kr->token_index[i]);
+
+	// Find the end of the last token, so we get the overall length of
+	// token_table. Then copy the token_table into host memory.
+	last_token = loc->kallsyms_token_table + kr->token_index[UINT8_MAX];
+	do {
+		err = drgn_program_read_memory(prog, &data,
+					       last_token, 1, false);
+		if (err)
+			return err;
+
+		last_token++;
+	} while (data);
+	token_table_size = last_token - loc->kallsyms_token_table + 1;
+	kr->token_table = malloc(token_table_size);
+	if (!kr->token_table)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->token_table,
+				       loc->kallsyms_token_table,
+				       token_table_size, false);
+	if (err)
+		return err;
+
+	// Now find the end of the names array by skipping through it, then copy
+	// that into host memory.
+	names_idx = 0;
+	kr->long_names = guess_long_names(prog);
+	for (size_t i = 0; i < kr->num_syms; i++) {
+		err = drgn_program_read_u8(prog,
+					   loc->kallsyms_names + names_idx,
+					   false, &len_u8);
+		if (err)
+			return err;
+		len = len_u8;
+		if ((len & 0x80) && kr->long_names) {
+			err = drgn_program_read_u8(prog,
+						loc->kallsyms_names + names_idx + 1,
+						false, &len_u8);
+			if (err)
+				return err;
+			len = (len & 0x7F) | (len_u8 << 7);
+			names_idx++;
+		}
+		names_idx += len + 1;
+	}
+	kr->names = malloc(names_idx);
+	if (!kr->names)
+		return &drgn_enomem;
+	err = drgn_program_read_memory(prog, kr->names,
+				       loc->kallsyms_names,
+				       names_idx, false);
+	if (err)
+		return err;
+
+	return NULL;
+}
+
+/**
+ * Extract the symbol name and type
+ * @param kr Registry containing kallsyms data
+ * @param offset Starting index within names array for this symbol
+ * @param sb Buffer to write output symbol to
+ * @param[out] kind_ret Where to write the symbol kind data
+ * @returns The offset of the next symbol in the names array
+ */
+static unsigned int
+kallsyms_expand_symbol(struct kallsyms_reader *kr, unsigned int offset,
+		       struct string_builder *sb, char *kind_ret)
+{
+	uint8_t *data = &kr->names[offset];
```

Comment:
It's *technically* possible that the data that we read when we read the whole table is different from the data that we read when we skipped through the lengths. (Yes, practically that won't happen, but imagine a buggy memory reader or something like that.) So, we need to make sure that `offset` is in bounds here and when we read the long length.

Likewise, check that `offset + len` is within bounds, too.